Sprint_15

Wilwous opened the pull request:
- Подключен Core Data для управления данными трекеров
- Добавлены классы для работы с Core Data: CoreDataStack, TrackerStore, TrackerCategoryStore, TrackerRecordStore
- Реализована поддержка хранения и управления трекерами, категориями и записями
- Добавлена коллекция цветов и эмодзи для создания трекера
- Произведен рефакторинг кода для переиспользования

Commit:
Добавил ассеты цветов и лого.
Сверстал лаунчскрин.

Commit:
feat: Добавлен TabBarViewController

- Добавлены вкладки "Трекеры" и "Статистика" в TabBarViewController
- Добавлена разделяющая линия

feat: Добавлен TrackerViewController

- Реализован поиск в TrackerViewController
- Добавлена возможность выбора даты
- Добавлена кнопка для добавления трекеров
- Добавлены лейблы для различных элементов в TrackerViewController

feat: Добавлен StatisticsViewController

chore: Добавлена иконка приложения и прочие ассеты.

Commit:
fix: Исправление и доработка форматирования даты

- Добавлены констрейны ширины для корректного отображения форматирования даты в цифрах
- Удалена дублирующая строка кода

Commit:
feat: Добавлены структуры Tracker, TrackerCategory и TrackerRecord

- Создана сущность Tracker для хранения информации о трекере с уникальным идентификатором, названием, цветом, эмоджи и расписанием.
- Введена TrackerCategory, представляющая категорию трекеров с заголовком и массивом трекеров внутри.
- Добавлена TrackerRecord для хранения записей о выполнении трекера на определенную дату с id трекера и датой выполнения.

feat: Обновлены свойства и методы в TrackersViewController

- Добавлены свойства categories и completedTrackers для хранения списка категорий и выполненных трекеров соответственно.
- Реализован механизм добавления новых трекеров в категории и отметок о выполнении в completedTrackers.

Commit:
fix

Commit:
feat: Добавлена функциональность кнопки добавления трекера

- Добавлена приватная функция addTrackerButtonTapped для обработки нажатия на кнопку добавления трекера.
- Реализован механизм перехода к экрану создания трекера при нажатии на кнопку.

Добавлен экран для создания трекеров

- Создан экран TrackerCreator для создания новых трекеров.
- Добавлены элементы интерфейса: метка "Создание трекера", кнопки "Привычка" и "Нерегулярное событие".
- Реализованы констрейнты для размещения элементов интерфейса на экране.

chore: Переименована переменная для лучшей читаемости

Commit:
UPD:
feat: Добавлена функциональность кнопки добавления трекера
- Добавлена приватная функция addTrackerButtonTapped для обработки нажатия на кнопку добавления трекера.
- Реализован механизм перехода к экрану создания трекера при нажатии на кнопку.

Добавлен экран для создания трекеров

- Создан экран TrackerCreator для создания новых трекеров.
- Добавлены элементы интерфейса: метка "Создание трекера", кнопки "Привычка" и "Нерегулярное событие".
- Реализованы констрейнты для размещения элементов интерфейса на экране.

chore: Переименована переменная для лучшей читаемости

Commit:
feat: Добавлен экран создания привычек

- Создан экран HabitCreation для создания новых привычек.
- Добавлены элементы интерфейса: метка "Новая привычка", текстовое поле для ввода названия привычки, сообщение об ограничении символов, таблица для выбора категории и расписания, кнопки "Отменить" и "Создать".
- Реализованы констрейнты и настройка интерфейса для корректного отображения элементов на экране.
- Реализована таблица с ячейками для выбора категории и расписания привычки.

chore: Переорганизация структуры проекта

Commit:
feat: Добавлен экран создания расписания

- Создан класс TimetableCreation для отображения экрана создания расписания.
- Реализован выбор дней недели с помощью UITableView.
- Добавлена кнопка "Готово" для завершения выбора дней недели.
- Добавлен протокол TimetableCreationDelegate для обратной связи с родительским контроллером.
- Создан класс TimetableTableView для ячейки таблицы с днями недели.
- Добавлена возможность выбора дней недели с помощью UISwitch в ячейке таблицы.

Commit:
feat: Добавлена коллекция трекера.
- Реализовано отображение заглушки.
- Добавлена возможность отметки трекера для пометки выполненных задач.
- Реализовано всплывающее сообщение о ограничение на количество символов.
- Добавлена функциональность фильтрации по дате и поиску для удобства работы с данными.

refactor:
- Переименованы переменные для более ясного понимания их назначения.
- Добавлены комментарии Mark для улучшения навигации и структурирования кода.
- Организован код и файлы для улучшения читаемости и поддерживаемости проекта.
- Пересмотрены и скорректированы констрейнты для улучшения внешнего вида интерфейса.

Commit:
feat: Добавлена функциональность создания и отмены трекера.
Введен алерт для информирования пользователя об ошибке при отсутствии названия трекера.
Реализована смена состояния кнопки при готовности создания трекера.

Commit:
fix: Поправил смену состояние кнопки создать.

Commit:
fix: - Устранена проблема с нажатием по таблице в выборе расписания трекера
- Исправлена проблема с ассетами: теперь появляется галочка завершения трекера и заглушка при отсутствии трекера
- Исправлена проблема, при которой страница создания трекера открывалась сразу
- Улучшена фильтрация поиска по выбранному дню
- Исправлено расположение кнопок создания и отмены трекера
- Улучшен функционал создания нового трекера

refactor: - Заменен поиск трекера с использованием UISearchController на UISearchBar
- Проведена реструктуризация кода для сокращения его объема

Commit:
fix: - Поправил ассеты
- Поправил нажатие на таблицу

Commit:
refactor: Убрал лишний код

Commit:
feature: - Создан экран нерегулярного события
- Добавлена коллекция цветов и эмодзи для создания трекера

refactor:
- Произведен рефакторинг кода для переиспользования

fix:
- Исправлен цвет лейбла трекера в соответствии с макетом

Commit:
feature: - Подключен Core Data для управления данными трекеров
- Добавлены классы для работы с Core Data: CoreDataStack, TrackerStore, TrackerCategoryStore, TrackerRecordStore
- Реализована поддержка хранения и управления трекерами, категориями и записями

So1ovyov commented on Tracker/CoreDateService/CoreDataStack.swift:
```diff
@@ -0,0 +1,41 @@
+import Foundation
+import CoreData
+
+final class CoreDataStack {
```
Хорошее решение

So1ovyov commented on Tracker/CoreDateService/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,112 @@
+    
+    // MARK: - Private Properties
+    private let managedObjectContext: NSManagedObjectContext
+    private var fetchedResultsController: NSFetchedResultsController<TrackerCoreData>!
```
В ТЗ требование "Нет ошибочных и грубых force-операций", давай сделаем опциональное свойство.
```
private var fetchedResultsController: NSFetchedResultsController<TrackerCoreData>?
```

So1ovyov commented on Tracker/CoreDateService/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,112 @@
+            return results.first
+        } catch {
+            print("Failed to fetch category by name: \(error)")
+            return nil
```
Это касается всех методов в файлах типа `Store`, давай вместо опционалов и пустых массивов обрабатывать ошибку через `throws`
```
/// Создаем ошибки для каждого Store

enum TrackerCategoryStoreError: Error {
    case fetchCategoryByNameError
    case fetchCategoryByNameNoValueError
}

/// Обрабатываем ошибки в методах

    func fetchCategoryByName(name: String) throws -> TrackerCategoryCoreData {
        let fetchRequest: NSFetchRequest<TrackerCategoryCoreData> = TrackerCategoryCoreData.fetchRequest()
        fetchRequest.predicate = NSPredicate(format: "title == %@", name)
        
        do {
            let results = try managedObjectContext.fetch(fetchRequest)
            guard let first = results.first else {
                throw TrackerCategoryStoreError.fetchCategoryByNameNoValueError
            }
            return first
        } catch {
            throw TrackerCategoryStoreError.fetchCategoryByNameError
        }
    }
```

So1ovyov replied:
Далее по цепочку ты сможешь во ViewController отлавливать ошибки и обрабатывать их

So1ovyov commented on Tracker/CoreDateService/TrackerRecordStore.swift:
```diff
@@ -0,0 +1,117 @@
+    
+    // MARK: - Private Properties
+    private let managedObjectContext: NSManagedObjectContext
+    private var fetchedResultsController: NSFetchedResultsController<TrackerCoreData>!
```
Аналогично и здесь
```
private var fetchedResultsController: NSFetchedResultsController<TrackerCoreData>?
```

So1ovyov commented on Tracker/CoreDateService/TrackerStore.swift:
```diff
@@ -0,0 +1,129 @@
+    
+    // MARK: - Private Properties
+    private let managedObjectContext: NSManagedObjectContext
+    private var fetchedResultsController: NSFetchedResultsController<TrackerCoreData>!
```
И не забудь здесь

So1ovyov commented on Tracker/CoreDateService/TrackerStore.swift:
```diff
@@ -0,0 +1,129 @@
+        let trackerCoreData = TrackerCoreData(context: managedObjectContext)
+        trackerCoreData.id = tracker.id
+        trackerCoreData.name = tracker.name
+        //            trackerCoreData.color = tracker.color.toHexString()
```
Закомментированный код можно удалить, если он не используется

So1ovyov commented on Tracker/CoreDateService/TrackerStore.swift:
```diff
@@ -0,0 +1,129 @@
+        trackerCoreData.timetable = tracker.timetableToJSON()
+        trackerCoreData.category = category
+        
+        print("🍔 Saving tracker with name: \(tracker.name), category: \(category.title ?? "Unknown")")
```
Print для дебага давай удалим

So1ovyov commented on Tracker/Helpers/CodableColor.swift:
```diff
@@ -0,0 +1,32 @@
+
+struct CodableColor: Codable {
+    
+    let color: UIColor
```
Если `color` задается через инициализатор, то почему свойство не пометить как `private`
```
private let color: UIColor
```

So1ovyov commented on Tracker/Helpers/DataSource.swift:
```diff
@@ -0,0 +1,35 @@
+
+import UIKit
+
+final class DataSource {
```
Возможно в прошлой итерации говорил об этом, но класс, а тем более синглтон это не самое лучшее решение для моковых данных. Лучше использовать `enum DataSource` со статическим свойством 
```
static let trackerCategories: [TrackerCategory] = [data]
```

So1ovyov commented on Tracker/Helpers/UIColor + Extension.swift:
```diff
@@ -0,0 +1,51 @@
+
+import UIKit
+
+extension UIColor {
```
Рекомендую всегда использовать библиотеки генерации ресурсов типа [SwiftGen](https://github.com/SwiftGen/SwiftGen).
Избавит от многих проблем при работе с ресурсами(цветами, строками, картинками и прочим).

So1ovyov commented on Tracker/Models/Tracker.swift:
```diff
@@ -0,0 +1,54 @@
+struct Tracker: Codable {
+    let id: UUID
+    let name: String
+//    🥎
```
Кажется это лишнее

So1ovyov commented on Tracker/TabBarManager/StatisticsManager/StatisticsViewController.swift:
```diff
@@ -0,0 +1,77 @@
+        [titleLabel,
+         emptyStateImageView,
+         emptyStateLabel
+        ].forEach {
```
Отличный инструмент !

So1ovyov commented on Tracker/TrackerManager/CollectionView/EmojiCollection.swift:
```diff
@@ -0,0 +1,57 @@
+    static let idetnifier = "EmojiCollection"
+    
+    // MARK: - Private Properties
+    lazy var emojiLabel: UILabel = {
```
Перепроверь все свойства и методы в классах, они должны быть по максимуму инкапсулированы и скрыты внутри класса. Используй модификатор `private`. Для того, чтобы достучаться до свойств, используй методы для настройки, но не как сейчас напрямую через открытые свойства.

So1ovyov commented on Tracker/TrackerManager/Habit/HabitTableView.swift:
```diff
@@ -0,0 +1,122 @@
+        return title
+    }()
+    
+    let subtitleLabel: UILabel = {
```
Аналогично и здесь, свойство `private`, а чтобы его обновить создать метод:
```
func updateSubtitle(text: String) {
   subtitleLabel.text = text
}
```

So1ovyov commented on Tracker/TrackerManager/View/TrackerHeader.swift:
```diff
@@ -0,0 +1,67 @@
+    static let headerIdentifier = String(describing: TrackerHeader.self)
+    
+    // MARK: - Public Properties
+    lazy var titleLabel: UILabel = {
```
Проверь все UI элементы в классах, инкапсулировать их очень важно.

So1ovyov commented on Tracker/AppDelegate.swift:
```diff
@@ -6,14 +6,21 @@
 //
 
 import UIKit
+import CoreData
```
В классе ничего не используется из `CoreData`, следовательно `import` тоже не нужен.

Commit:
feature: - Добавлено создание нерегулярного события с сохранением в CoreData

refactor:
- Убран скролл эмодзи и цветов в создании трекера
- Сделан fetchedResultsController опциональным
- Проведена корректировка кода

fix:
- Исправлен цвет лейбла трекера в соответствии с макетом

Wilwous commented:
- Добавлено создание нерегулярного события с сохранением в CoreData
- Убран скролл эмодзи и цветов в создании трекера
- Сделан fetchedResultsController опциональным
- Проведена корректировка кода
- Исправлен цвет лейбла трекера в соответствии с макетом

Wilwous commented:
Критические замечания исправлены. 
@So1ovyov Благодарю за рекомендации и советы по коду, они будут учтены и исправлены в дальнейших спринтах или по завершению модуля.
